diem-move/diem-framework/experimental/sources/NFT.move

```diff
@@ -1,93 +1,264 @@
+        public fun parent<TokenType: store>(wrapper: &TokenDataWrapper<TokenType>): &Option<GUID::ID> {
+            &wrapper.parent_id
+        }
+
+        /// Returns the supply of tokens with `id` on the chain.
+        public fun supply<TokenType: store>(id: &GUID::ID): u64 acquires TokenDataCollection {
+            let owner_addr = GUID::id_creator_address(id);
+            let tokens = &mut borrow_global_mut<TokenDataCollection<TokenType>>(owner_addr).tokens;
+            let index_opt = index_of_token<TokenType>(tokens, id);
+            assert!(Option::is_some(&index_opt), Errors::invalid_argument(EWRONG_TOKEN_ID));
+            let index = Option::extract(&mut index_opt);
+            Vector::borrow(tokens, index).supply
+        }
+
+        /// Extract the Token data of the given token into a hot potato wrapper.
+        public fun extract_token<TokenType: store>(nft: &Token<TokenType>): TokenDataWrapper<TokenType> acquires TokenDataCollection {
+            let owner_addr = GUID::id_creator_address(&nft.id);
+            let tokens = &mut borrow_global_mut<TokenDataCollection<TokenType>>(owner_addr).tokens;
+            let index_opt = index_of_token<TokenType>(tokens, &nft.id);
+            assert!(Option::is_some(&index_opt), Errors::invalid_argument(EWRONG_TOKEN_ID));
+            let index = Option::extract(&mut index_opt);
+            let item_opt = &mut Vector::borrow_mut(tokens, index).metadata;
+            assert!(Option::is_some(item_opt), Errors::invalid_state(ETOKEN_EXTRACTED));
+            let metadata = Option::extract(item_opt);
+            let parent_opt = &mut Vector::borrow_mut(tokens, index).parent_id;
+            TokenDataWrapper { origin: owner_addr, index, metadata, parent_id: *parent_opt }
+        }
+
+        /// Restore the token in the wrapper back into global storage under original address.
+        public fun restore_token<TokenType: store>(wrapper: TokenDataWrapper<TokenType>) acquires TokenDataCollection {
+            let TokenDataWrapper { origin, index, metadata, parent_id: _ } = wrapper;
+            let tokens = &mut borrow_global_mut<TokenDataCollection<TokenType>>(origin).tokens;
+            assert!(Vector::length(tokens) > index, EINDEX_EXCEEDS_LENGTH);
+            let item_opt = &mut Vector::borrow_mut(tokens, index).metadata;
+            assert!(Option::is_none(item_opt), ETOKEN_PRESENT);
+            Option::fill(item_opt, metadata);
+        }
+
+        /// Finds the index of token with the given id in the gallery.
+        fun index_of_token<TokenType: store>(gallery: &vector<TokenData<TokenType>>, id: &GUID::ID): Option<u64> {
+            let i = 0;
+            let len = Vector::length(gallery);
+            while (i < len) {
+                if (GUID::eq_id(&Vector::borrow(gallery, i).token_id, id)) {
+                    return Option::some(i)
+                };
+                i = i + 1;
+            };
+            Option::none()
+        }
```

Comment:
There's a lot of duplication in here.  We should later (not in this PR), consider either removing the deduplication, or asking the move team to provide these kinds of functionalities (e.g. iterate to find token or even just for loops)